### PR TITLE
Implement Merge Sort

### PR DESCRIPTION
**Pull request by TravorLZH**
I implemented merge sort for both integers and generic types. (#10)

### CODE REVIEW

**Comment by ParadoxZero** [comment]
Rebase your code once with the updated master

**Comment by ParadoxZero** [comment]
There seems to be some issue with your branch. The changes made haven't registered e.g. `binary_tree.h` `queue.h`.

Also, the test are failing. Reason - Not being run.

**Comment by TravorLZH** [comment]
I put binaries in `build/` folder, so you need to change the test location

File: src/data_structure/binary_tree.c [hunk 1]
```diff
@@ -0,0 +1,109 @@
+/*
```
Comment (ParadoxZero) [inline src/data_structure/binary_tree.c:1]: remove this file, it has been renamed in master.

File: include/generic/list.h [hunk 2]
```diff
@@ -37,7 +37,7 @@
 #include <generic.h>
 #include <stddef.h>
 #include <stdlib.h>
-
+#include <inttypes.h>
```
Comment (ParadoxZero) [inline include/generic/list.h]: revert this file

File: include/data_structure/queue.h [hunk 3]
```diff
@@ -0,0 +1,85 @@
+/*
```
Comment (ParadoxZero) [inline include/data_structure/queue.h:1]: remove.

I have renamed it to `vqueue` in the master.

Why didn't the change reflect when you rebased/merged? 😕

File: include/data_structure/binary_tree.h [hunk 4]
```diff
@@ -0,0 +1,106 @@
+/*
```
Comment (ParadoxZero) [inline include/data_structure/binary_tree.h:1]: Same as the case of `vqueue` Renamed this file in master.

File: include/algorithm/merge_sort.h [hunk 5]
```diff
@@ -0,0 +1,70 @@
+/*
```
Comment (ParadoxZero) [inline include/algorithm/merge_sort.h:1]: Rename file to `mergesort.h`

Follow the code style specified in [wiki](https://github.com/ParadoxZero/libgeneric/wiki/Code-guide)

File: include/algorithm/algorithm.h [hunk 6]
```diff
@@ -0,0 +1,39 @@
+/*
```
Comment (ParadoxZero) [inline include/algorithm/algorithm.h:1]: defined similar type in `utils.h`, use it instead of declaring new.

File: Test/btree/btree_search.c [hunk 7]
```diff
@@ -0,0 +1,25 @@
+#include <data_structure/binary_tree.h>
```
Comment (ParadoxZero) [inline Test/btree/btree_search.c:1]: Master has changes to this file.

File: Test/vector/vector_stack.c [hunk 8]
```diff
@@ -0,0 +1,34 @@
+#include <data_structure/vector.h>
```
Comment (ParadoxZero) [inline Test/vector/vector_stack.c:1]: not a test.

Implement it as `vstack`.

File: Test/btree/btree_print.c [hunk 9]
```diff
@@ -0,0 +1,18 @@
+#include <data_structure/binary_tree.h>
```
Comment (ParadoxZero) [inline Test/btree/btree_print.c:1]: Removed this function and test in master.

File: Test/algorithm/mergesort_test.c [hunk 10]
```diff
@@ -0,0 +1,27 @@
+#include <stdio.h>
```
Comment (ParadoxZero) [inline Test/algorithm/mergesort_test.c:1]: remove

File: src/algorithm/merge_sort.c [hunk 11]
```diff
@@ -0,0 +1,140 @@
+/*
```
Comment (ParadoxZero) [inline src/algorithm/merge_sort.c:1]: rename to `mergesort.c`

Comment (ParadoxZero) [reply src/algorithm/merge_sort.c:1]: Also follow the code guide.

**Review by ParadoxZero — CHANGES_REQUESTED** [review]
Are you sure you rebased correctly?

the history of master branch is not reflected in this. Try to correct the issue.

Moreover, I have commented few changes please go through it.

Please go through the code guide, and follow it.

File: src/algorithm/merge_sort.c (hunk 11 above)
Comment (ParadoxZero) [reply src/algorithm/merge_sort.c:1]: Only generic merge sort needed. No need to special int.

**Review by ParadoxZero — COMMENTED** [review]
Instead of merging, try to rebase. Maybe it would fix the issue.

**Comment by ParadoxZero** [comment]
Tests working, Now all that's left is to resolve the changes I mentioned.

**Comment by ParadoxZero** [comment]
The git history of this branch is a mess.

Ensure the only files with changes should be:

- `include/genetic/algorithm.h` Here the declaration of mergesort should exist.
- `src/algorithm/mergesort.c` Here the source of mergesort.
- `Test/algorithm/test_mergesort.c` (the name is generic) The test for merge sort.
- `Test/CMakeList.txt` to add algorithms to test suit.
- `Test/algorithm/CMakeList.txt` to add mergesort test to test suit.

Optionally,
- `Readme.md` But the only change should be that mergesort be ticked off.

Clean the branch history and ensure this order.

Only then will I be able to merge the request.

**Pull request closed without merging**